crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -55,6 +63,84 @@ impl fmt::Debug for RoomEventCache {
     }
 }
 
+/// A RAII data structure that will automatically shrink the current room's
+/// linked chunk on drop.
+pub(super) struct AutoShrinkLinkedChunk {
+    /// The related room event cache, wrapped in an option so as to take it from
+    /// the drop implementation.
+    room_event_cache: Option<RoomEventCache>,
+}
+
+impl Drop for AutoShrinkLinkedChunk {
+    fn drop(&mut self) {
+        let room_event_cache = self.room_event_cache.take().expect("drop called multiple times");
+
+        // I hear you from the future: "but, spawning a detached task in a drop
+        // implementation is real bad! Maybe there will be multiple shrinks
+        // happening at the same time, and that's bad!". However, this can't
+        // happen, because the whole `state` variable is guarded by a fair lock,
+        // which will run queries in the order they happen. Should be fine™.
```

Comment:
You forgot to explain why you need to `spawn` here. It's because `drop` cannot be async, and you're doing async operations.

Reply:
Yep, introduced a mechanism with one (1) task listening forever in the event cache, that should be more sound.